### PR TITLE
Add System.CodeDom tests

### PR DESCRIPTION
**Pull request by hughbe**
All of the "model" stuff
Next up, the collections
After, csharp specific generation stuff

Contributes to #12179
/cc @stephentoub

### CODE REVIEW

**Comment by hughbe** [comment]
Oh wow, didn't realise this one ended up at 5,000 lines.
Thankfully, there are really phew deletions, so that should make things a bit easier!

File: src/System.CodeDom/tests/Other/CodeCatchClauseTests.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    public class CodeCatchClauseTests : CodeDomTestBase
+    {
+        [Fact]
+        public void Ctor_Default()
+        {
+            var catchClause = new CodeCatchClause();
+            Assert.Empty(catchClause.LocalName);
+            Assert.Empty(catchClause.Statements);
+
+			Assert.Equal(new CodeTypeReference(typeof(Exception)).BaseType, catchClause.CatchExceptionType.BaseType);
```
Comment (justinvp) [inline src/System.CodeDom/tests/Other/CodeCatchClauseTests.cs]: Nit: Indentation in this method looks inconsistent.

**Comment by stephentoub** [comment]
@dotnet-bot test code coverage please

File: src/System.CodeDom/tests/CodeExpressions/CodeDelegateInvokeExpressionTests.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+			Assert.Throws<ArgumentNullException>("value", () => new CodeDelegateInvokeExpression(new CodePrimitiveExpression("Hello"), parameters));
+		}
+
+		[Theory]
+		[MemberData(nameof(CodeExpression_TestData))]
+		public void TargetObject_Set_Get_ReturnsExpected(CodeExpression value)
+		{
+			var delegateInvoke = new CodeDelegateInvokeExpression();
+			delegateInvoke.TargetObject = value;
+			Assert.Equal(value, delegateInvoke.TargetObject);
```
Comment (stephentoub) [inline src/System.CodeDom/tests/CodeExpressions/CodeDelegateInvokeExpressionTests.cs]: Nit: Assert.Same?

File: src/System.CodeDom/tests/CodeGenerationTests.cs [hunk 3]
```diff
@@ -191,4 +192,15 @@ private static string CoalesceWhitespace(string str)
             return sb.ToString();
         }
     }
+
+    public static class Common
+    {
+        public static IEnumerable<T> ToIEnumerable<T>(this IEnumerator<T> enumerator)
```
Comment (stephentoub) [inline src/System.CodeDom/tests/CodeGenerationTests.cs]: Where is this used? I assume it's here for your upcoming collection tests or something?

**Comment by stephentoub** [comment]
This is excellent, @hughbe.  Thank you for doing all this!  A few minor comments you can choose to address or not, and some strange formatting throughout that looks like it may be a mix of tabs and spaces, but otherwise this is terrific.

**Comment by stephentoub** [comment]
(It's great seeing all of those 100%s in the CodeDom coverage report now :smile: https://ci.dot.net/job/dotnet_corefx/job/master/job/code_coverage_windows_prtest/2/artifact/bin/tests/coverage/index.htm)

**Comment by hughbe** [comment]
Thanks, it seems VS synchronizes my tabs and spaces settings between the machines at home and work, sorry about that!

**Comment by stephentoub** [comment]
@dotnet-bot Test Innerloop CentOS7.1 Release Build and Test please
```
Initializing BuildTools...
ERROR: An error occured when trying to initialize the tools. Please check '/mnt/resource/j/workspace/dotnet_corefx/master/centos7.1_release_prtest/init-tools.log' for more details.1
```
cc: @weshaggard

**Pull request merged**